contributors/design-proposals/network/service-topology-key.md

```diff
@@ -0,0 +1,224 @@
+
+```go
+type ServicePolicy struct {
+  TypeMeta
+  ObjectMeta
+  Spec ServicePolicySpec
+}
+
+type ServicePolicySpec struct {
+  // ServiceSelector chooses what services the Service Policies should apply.
+  ServiceSelector *metav1.LabelSelector `json:"endPointSelector" protobuf:"bytes,1,opt,name=podSelector"`
+
+  // topology is one of the service policies, which is used to achieve "local" service in a given topology level.
+  // Cluster administrators can configure what ever topological preferences they want.
+  // +optional
+  Topology ServiceTopology `json:"topology" protobuf:"bytes,2,opt,name=topology"`
+}
+
+// Defines a service topolgoy information.
+type ServiceTopology struct {
+  // Valid values for mode are "ignored", "required", "preferred".
+  // "ignored" is the default value and the associated topology key will have no effect.
+  // "required" is the "hard" requirement for topology key and an example would be  “only visit service backends in the same zone”.
+  // If the topology requirements specified by this field are not met, the LB, such as kube-proxy will not pick endpoints for the service.
+  // "preferred" is the "soft" requirement for topology key and an example would be
+  // "prefer to visit service backends in the same rack, but OK to other racks if none match"
+  // +optional
+  Mode ServicetopologyMode `json:"mode" protobuf:"bytes,1,opt,name=mode"`
+
+  // key is the key for the node label that the system uses to denote
+  // such a topology domain. There are some built-in topology keys, e.g.
+  // kubernetes.io/hostname, failure-domain.beta.kubernetes.io/zone and failure-domain.beta.kubernetes.io/region etc.
+  // The built-in topology keys can be good examples and we recommend users switch to a similar mode for portability, but it's NOT enforced.
+  // Users can define whatever topolgoy key they like since topology is arbitrary.
+  // +optional
+  Key string `json:"key" protobuf:"bytes,2,opt,name=key"`
+}
+```
+
+An example of `ServicePolicy` might looks like,
+
+```yaml
+kind: ServicePolicy
+metadata:
+  name: service-policy-example
+  namespace: foo
+spec:
+  serviceSelector:
+    matchLabels:
+      app: bar
+  topology:
+    key: kubernetes.io/hostname
+    mode: required
+```
+
+In our example, services in namespace `foo` with label `app=bar` will be chosen. Requests to these services will be routed only to backends on nodes with the same value for `kubernetes.io/hostname` as the originating pod's node. If we want the "same host", probably every host should carry unique `kubernetes.io/hostname` labels.
+
+We can configure multiple `ServicePolicy` targeting a single service. In this case, the service will carry multiple topology requirements and the relationships of all the requirements are logical `AND`, for example,
+
+```yaml
+kind: ServicePolicy
+metadata:
+  name: service-policy-example-1
+  namespace: foo
```

Comment:
@thockin - does this approach scale when having a lot of namespaces ? can we use a namespaceSelector ? or a list namespaces ? and make this a non-namespaced resource ?